Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift

```diff
@@ -523,26 +524,31 @@ struct PathHierarchy {
                     return possibleMatches.first(where: { $0.symbol?.identifier.interfaceLanguage == "swift" }) ?? possibleMatches.first!
                 }
                 // Couldn't resolve the collision by look ahead.
-                return try handleCollision(node: node, parsedPath: parsedPathForError(), remaining: remaining, collisions: collisions)
+                return try handleCollision(node: node, parsedPath: parsedPathForError, remaining: remaining, collisions: collisions, onlyFindSymbols: onlyFindSymbols)
             }
         }
     }
                         
     private func handleCollision(
         node: Node,
-        parsedPath: [PathComponent],
+        parsedPath: () -> [PathComponent],
         remaining: ArraySlice<PathComponent>,
-        collisions: [(node: PathHierarchy.Node, disambiguation: String)]
+        collisions: [(node: PathHierarchy.Node, disambiguation: String)],
+        onlyFindSymbols: Bool
     ) throws -> Node {
-        let favoredNodes = collisions.filter { $0.node.isDisfavoredInCollision == false }
-        if favoredNodes.count == 1 {
-            return favoredNodes.first!.node
+        if let favoredMatch = collisions.singleMatch({ $0.node.isDisfavoredInCollision == false }) {
+            return favoredMatch.node
+        }
+        // If non-symbols are allowed, those results are preferred over any symbols. It's recommended to use symbol links to find symbols.
+        // If non-symbols aren't allowed, skip all non-symbols.
+        if let symbolOrNonSymbolMatch = collisions.singleMatch({ ($0.node.symbol != nil) == onlyFindSymbols }) {
+            return symbolOrNonSymbolMatch.node
```

Comment:
I'm struggling to follow this line. Could it be structured like this or is it doing something different?

```suggestion
        if !onlyFindSymbols, let nonSymbolMatch = (collisions.singleMatch { $0.node.symbol == nil })
            return nonSymbolMatch.node
```

> If non-symbols aren't allowed, skip all non-symbols.

Have these not already been filtered out prior to this or do we need to handle that case here?

Reply:
The only difference is that with `onlyFindSymbols` inside the closure it checks both ways. 

For symbol links (when `onlyFindSymbols` is `true`) it checks that the match has a symbol.
For general documentation links (when `onlyFindSymbols` is `false`) it checks that the match doesn't have a symbol.

> Have these not already been filtered out prior to this or do we need to handle that case here?

It is checked at the start of the search to avoid looking at the article root but as I was writing this I thought that it would be possible to find an article as a child of the module. Now I'm not sure anymore.. I'll have to try it.

Reply:
Okay sounds good. If it's possible to simplify this I think that would make it easier to follow – but not a blocker.

Reply:
I tested this and in the case where a symbol and an article have the same path, a link that's spelled with the absolute path will have a collisions here. In this case filtering out non-symbols in symbol links in both places (here and where the found node is returned) provides a better experience because it avoids the collisions (and also excludes the non-symbols from diagnostics if there were multiple symbols colliding with an article).

I updated the comment to explain all this, added a test that has this behavior, and added an extra comment that explains that check is a more compact spelling of two comparisons and an if statement.